### PR TITLE
[WIP] Bidirectional branching

### PR DESCRIPTION
**Pull request by zovt**
#### Summary of changes
This is a proof-of-concept PR to demonstrate bidirectional branching support in the line diagram rendered next to the stop information. 

The main conceptual change here is adjusting `branchingOutward` from being a page/line-global decision to a per-stop calculation based on the surrounding data in the stops list. I've tried to keep the rest of the code as unmodified as possible to focus on this, though there are definitely some areas to clean up. The only other changes were things that were broken when a branch was small enough not to have collapsed stops.

As it stands, it works for the data that I've been testing with (which is included in the diff), though I would hesitate to call this "complete" as the code still feels pretty edgy and would definitely like to test this with real(er?) data. I've experimented with moving the test branch around in the stop list and things appear to work properly, and other lines render properly as well.

### CODE REVIEW

File: apps/site/assets/ts/schedule/components/line-diagram/LineDiagramWithStops.tsx [hunk 1]
```diff
@@ -15,7 +16,57 @@ export const StopRefContext = React.createContext<[RefList, () => void]>([
...
+      ], { "is_terminus?": true, branch: "Green Test Split", }),
+      fake("Test Station C", [
+          { "has_disruption?": false, branch: null, type: 'stop' },
+      ]),
+    ];
+    stops = stops_[0].route_stop.name === 'North Station'
+      ? [...extraA, ...stops_]
+      : [...stops_, ...extraB];
+    console.debug({stops});
+  }
```
[the remark this answers is not in the document]
Comment (thecristen) [reply apps/site/assets/ts/schedule/components/line-diagram/LineDiagramWithStops.tsx:69]: We'll have to make sure the backend can output every stop with the correct `stop_data` structure, branch naming, terminus labelling, etc.

**Review by thecristen — COMMENTED** [review]
I'm not sure about how well it would handle infinite branches on branches in any directions — but this seems like a solid approach in the frontend to handling the bidirectional branching case that we anticipate for the combined Green line diagram once GLX is complete. I'm frankly impressed with how light of a modification was needed for that.

We're still missing the backend piece, with is crucial as it supplies the frontend with the stop information. But at least with this in place, we can rest assured that the backend won't have to output anything dramatically different than before to render a decent diagram.

**Comment by zovt** [comment]
> I'm not sure about how well it would handle infinite branches on branches in any directions

I think it would not work at all :) 

> We're still missing the backend piece, which is crucial as it supplies the frontend with the stop information

Ah I see, I didn't realize that the backend needed any special handling for the directions of the branching.

**Comment by zovt** [comment]
For what it's worth I still think a longer-term solution to getting the branching "right" is to use a different data representation for the stops, as the flat list quickly becomes tricky to work with and understand for more complex cases.

**Comment by thecristen** [comment]
> Ah I see, I didn't realize that the backend needed any special handling for the directions of the branching.

Well, it's more like, we can't automatically expect North Station to have the expected :merge values in its `stop_data`. Maybe.

We now more or less tell the Green line branches how to coincide/overlap, and that logic probably needs adjusting to tell it run the N Station - Union & N Station - Medford branches to be side-by-side (by specifying the correct stop_data) AND have the prior B/C/D branches on the other end work as expected, within in the same list of stops.

As of now there's a fair amount of manual fiddling (overriding the logic and just hard-coding what each stop should do) for the northern bits of the Green line. Which itself could use improving or replacing.

**Comment by thecristen** [comment]
Closing because I'm finally adding Asana tickets to carry this out, thank you!

**Pull request closed without merging**